logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -301,13 +310,77 @@ private void deleteExpiredSegments() throws IOException {
     }
 
-    private void deleteTailSegment(Path segment, String motivation) throws IOException {
+    /**
+     * Deletes the segment path, if present. Also return the number of events it contains.
+     *
+     * @param segment
+     *      The segment file to delete.
+     * @param motivation
+     *      Description of delete motivation.
+     * @return the number of events contained in the segment or empty optional if the segment was already
+     *      removed.
+     * */
+    private Optional<Long> deleteTailSegment(Path segment, String motivation) throws IOException {
         try {
+            long eventsInSegment = countEventsInSegment(segment);
             Files.delete(segment);
             logger.debug("Removed segment file {} due to {}", motivation, segment);
+            return Optional.of(eventsInSegment);
         } catch (NoSuchFileException nsfex) {
             // the last segment was deleted by another process, maybe the reader that's cleaning consumed segments
             logger.debug("File not found {}, maybe removed by the reader pipeline", segment);
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Count the number of 'c' and 's' records in segment.
+     * An event can't be bigger than the segments so in case of records split across multiple event blocks,
+     * the segment has to contain both the start 's' record, all the middle 'm' up to the end 'e' records.
+     * */
+    @SuppressWarnings("fallthrough")
+    private long countEventsInSegment(Path segment) throws IOException {
+        try (FileChannel channel = FileChannel.open(segment, StandardOpenOption.READ)) {
+            // verify minimal segment size
+            if (channel.size() < VERSION_SIZE + RECORD_HEADER_SIZE) {
+                return 0L;
+            }
+
+            // skip the DLQ version byte
+            channel.position(1);
+            int posInBlock = 0;
+            int currentBlockIdx = 0;
+            long countedEvents = 0;
+            do {
+                ByteBuffer headerBuffer = ByteBuffer.allocate(RECORD_HEADER_SIZE);
+                long startPosition = channel.position();
+                // if record header can't be fully contained in the block, align to the next
+                if (posInBlock + RECORD_HEADER_SIZE + 1 > BLOCK_SIZE) {
+                    channel.position((++currentBlockIdx) * BLOCK_SIZE + VERSION_SIZE);
+                    posInBlock = 0;
+                }
+
+                channel.read(headerBuffer);
+                headerBuffer.flip();
+                RecordHeader recordHeader = RecordHeader.get(headerBuffer);
+                if (recordHeader == null) {
+                    logger.error("Can't decode record header, position {} current post {} current events count {}", startPosition, channel.position(), countedEvents);
+                    throw new IllegalStateException("Can't decode record header at position " + startPosition);
```

Comment:
I thought to throw the exception because the impossibility to recognize which record type is from the first byte of the record, is symptom of corrupted file or programming error.
- in the first case we could just log the error and move to the next record.
- in the second there isn't much we can do, just write a test to cover the use case and fix the code.

Do you think we should jump to the next record and log the error?

Reply:
Fixed  by e1458e0006adbdc4360c31343b5942e791037456